// electron-dev-console/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var electron_1 = require("electron");
var SIGNAL = 'ELECTRON_DEV_CONSOLE';
exports.main = {
    window: null,
    init: function (window) {
        exports.main.window = window;
    },
    info: function (msg) {
        exports.main.window.webContents.send(SIGNAL, 'info', msg);
    },
    success: function (msg) {
        exports.main.window.webContents.send(SIGNAL, 'success', msg);
    },
    warn: function (msg) {
        exports.main.window.webContents.send(SIGNAL, 'warn', msg);
    },
    error: function (msg) {
        exports.main.window.webContents.send(SIGNAL, 'error', msg);
    },
};
exports.renderer = {
    init: function () {
        var prefix = '%c electron-dev-console';
        var successPrefixStyle = 'color: #fff;background-color: #64f898; font-size: 16px;';
        var successMsgStyle = 'color: #64f898; font-size: 14px;';
        var warnPrefixStyle = 'color: #fff;background-color: #f4faa7; font-size: 16px';
        var warnMsgStyle = 'color: #f4faa7; font-size: 14px';
        var infoPrefixStyle = 'color: #fff;background-color: #62ceff; font-size: 16px';
        var infoMsgStyle = 'color: #62ceff; font-size: 14px';
        var errorPrefixStyle = 'color: #fff;background-color: #ff6661; font-size: 16px';
        var errorMsgStyle = 'color: #ff6661; font-size: 14px';
        electron_1.ipcRenderer.on(SIGNAL, function (e, type, msg) {
            switch (type) {
                case 'success': {
                    console.log(prefix, successPrefixStyle);
                    console.log("%c " + msg, successMsgStyle);
                    break;
                }
                case 'warn': {
                    console.log(prefix, warnPrefixStyle);
                    console.log("%c " + msg, warnMsgStyle);
                    break;
                }
                case 'info': {
                    console.log(prefix, infoPrefixStyle);
                    console.log("%c " + msg, infoMsgStyle);
                    break;
                }
                case 'error': {
                    console.log(prefix, errorPrefixStyle);
                    console.log("%c " + msg, errorMsgStyle);
                    break;
                }
                default: {
                    console.log(prefix, infoPrefixStyle);
                    console.log("%c " + msg, infoMsgStyle);
                }
            }
        });
    }
};
