// @wokwi/elements/dist/cjs/lcd1602-element.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LCD1602Element = void 0;
const lit_1 = require("lit");
const decorators_js_1 = require("lit/decorators.js");
const lcd1602_font_a00_1 = require("./lcd1602-font-a00");
const pin_1 = require("./pin");
const units_1 = require("./utils/units");
const charXSpacing = 3.55;
const charYSpacing = 5.95;
const backgroundColors = {
    green: '#6cb201',
    blue: '#000eff',
};
let LCD1602Element = class LCD1602Element extends lit_1.LitElement {
    constructor() {
        super(...arguments);
        this.color = 'black';
        this.background = 'green';
        this.characters = new Uint8Array(32);
        this.font = lcd1602_font_a00_1.fontA00;
        this.cursor = false;
        this.blink = false;
        this.cursorX = 0;
        this.cursorY = 0;
        this.backlight = true;
        this.pins = 'full';
        this.numCols = 16;
        this.numRows = 2;
    }
    get text() {
        return Array.from(this.characters)
            .map((c) => String.fromCharCode(c))
            .join('');
    }
    set text(value) {
        this.characters = new Uint8Array(value.split('').map((char) => char.charCodeAt(0)));
    }
    static get styles() {
        return lit_1.css `
      .cursor-blink {
        animation: cursor-blink;
      }

      @keyframes cursor-blink {
        from {
          opacity: 0;
        }
        25% {
          opacity: 1;
        }
        75% {
          opacity: 1;
        }
        to {
          opacity: 0;
        }
      }
    `;
    }
    get panelHeight() {
        return this.rows * 5.75;
    }
    get pinInfo() {
        const { panelHeight } = this;
        const y = 87.5 + panelHeight * units_1.mmToPix;
        switch (this.pins) {
            case 'i2c':
                return [
                    { name: 'GND', x: 4, y: 32, number: 1, signals: [{ type: 'power', signal: 'GND' }] },
                    { name: 'VCC', x: 4, y: 41.5, number: 2, signals: [{ type: 'power', signal: 'VCC' }] },
                    { name: 'SDA', x: 4, y: 51, number: 3, signals: [pin_1.i2c('SDA')] },
                    { name: 'SCL', x: 4, y: 60.5, number: 4, signals: [pin_1.i2c('SCL')] },
                ];
            case 'full':
            default:
                return [
                    { name: 'VSS', x: 32, y, number: 1, signals: [{ type: 'power', signal: 'GND' }] },
                    { name: 'VDD', x: 41.5, y, number: 2, signals: [{ type: 'power', signal: 'VCC' }] },
                    { name: 'V0', x: 51.5, y, number: 3, signals: [] },
                    { name: 'RS', x: 60.5, y, number: 4, signals: [] },
                    { name: 'RW', x: 70.5, y, number: 5, signals: [] },
                    { name: 'E', x: 80, y, number: 6, signals: [] },
                    { name: 'D0', x: 89.5, y, number: 7, signals: [] },
                    { name: 'D1', x: 99.5, y, number: 8, signals: [] },
                    { name: 'D2', x: 109, y, number: 9, signals: [] },
                    { name: 'D3', x: 118.5, y, number: 10, signals: [] },
                    { name: 'D4', x: 128, y, number: 11, signals: [] },
                    { name: 'D5', x: 137.5, y, number: 12, signals: [] },
                    { name: 'D6', x: 147, y, number: 13, signals: [] },
                    { name: 'D7', x: 156.5, y, number: 14, signals: [] },
                    { name: 'A', x: 166.5, y, number: 15, signals: [] },
                    { name: 'K', x: 176, y, number: 16, signals: [] },
                ];
        }
    }
    get cols() {
        return this.numCols;
    }
    get rows() {
        return this.numRows;
    }
    path(characters) {
        const xSpacing = 0.6;
        const ySpacing = 0.7;
        const result = [];
        const { cols } = this;
        for (let i = 0; i < characters.length; i++) {
            const charX = (i % cols) * charXSpacing;
            const charY = Math.floor(i / cols) * charYSpacing;
            for (let py = 0; py < 8; py++) {
                const row = this.font[characters[i] * 8 + py];
                for (let px = 0; px < 5; px++) {
                    if (row & (1 << px)) {
                        const x = (charX + px * xSpacing).toFixed(2);
                        const y = (charY + py * ySpacing).toFixed(2);
                        result.push(`M ${x} ${y}h0.55v0.65h-0.55Z`);
                    }
                }
            }
        }
        return result.join(' ');
    }
    renderCursor() {
        const { cols, rows, cursor, cursorX, cursorY, blink, color } = this;
        const xOffset = 12.45 + cursorX * charXSpacing;
        const yOffset = 12.55 + cursorY * charYSpacing;
        if (cursorX < 0 || cursorX >= cols || cursorY < 0 || cursorY >= rows) {
            return null;
        }
        const result = [];
        if (blink) {
            result.push(lit_1.svg `
        <rect x="${xOffset}" y="${yOffset}" width="2.95" height="5.55" fill="${color}">
          <animate
            attributeName="opacity"
            values="0;0;0;0;1;1;0;0;0;0"
            dur="1s"
            fill="freeze"
            repeatCount="indefinite"
          />
        </rect>
      `);
        }
        if (cursor) {
            const y = yOffset + 0.7 * 7;
            result.push(lit_1.svg `<rect x="${xOffset}" y="${y}" width="2.95" height="0.65" fill="${color}" />`);
        }
        return result;
    }
    renderI2CPins() {
        return lit_1.svg `
      <rect x="7.55" y="-2.5" height="2.5" width="10.16" fill="url(#pins)" transform="rotate(90)" />
      <text fill="white" font-size="1.5px" font-family= "monospace">
      <tspan y="6.8" x="0.7" fill="white">1</tspan>
      <tspan y="8.9" x="2.3" fill="white">GND</tspan>
      <tspan y="11.4" x="2.3" fill="white">VCC</tspan>
      <tspan y="14" x="2.3" fill="white">SDA</tspan>
      <tspan y="16.6" x="2.3" fill="white">SCL</tspan>
      </text>
    `;
    }
    renderPins(panelHeight) {
        const y = panelHeight + 21.1;
        return lit_1.svg `
      <g transform="translate(0, ${y})">
        <rect x="7.55" y="1" height="2.5" width="40.64" fill="url(#pins)" />
        <text fill="white" font-size="1.5px" font-family= "monospace">
          <tspan x="6" y="2.7">1</tspan>
          <tspan x="7.2" y="0.7">VSS</tspan>
          <tspan x="9.9" y="0.7">VDD</tspan>
          <tspan x="12.7" y="0.7">V0</tspan>
          <tspan x="15.2" y="0.7">RS</tspan>
          <tspan x="17.8" y="0.7">RW</tspan>
          <tspan x="20.8" y="0.7">E</tspan>
          <tspan x="22.7" y="0.7">D0</tspan>
          <tspan x="25.3" y="0.7">D1</tspan>
          <tspan x="27.9" y="0.7">D2</tspan>
          <tspan x="30.4" y="0.7">D3</tspan>
          <tspan x="33" y="0.7">D4</tspan>
          <tspan x="35.6" y="0.7">D5</tspan>
          <tspan x="38.2" y="0.7">D6</tspan>
          <tspan x="40.8" y="0.7">D7</tspan>
          <tspan x="43.6" y="0.7">A</tspan>
          <tspan x="46.2" y="0.7">K</tspan>
          <tspan x="48" y="2.7">16</tspan>
        </text>
      </g>
    `;
    }
    render() {
        const { color, characters, background, pins, panelHeight, cols } = this;
        const darken = this.backlight ? 0 : 0.5;
        const actualBgColor = background in backgroundColors ? backgroundColors[background] : backgroundColors;
        const panelWidth = cols * 3.5125;
        const width = panelWidth + 23.8;
        const height = panelHeight + 24.5;
        // Dimensions according to:
        // https://www.winstar.com.tw/products/character-lcd-display-module/16x2-lcd.html
        return lit_1.html `
      <svg
        width="${width}mm"
        height="${height}mm"
        version="1.1"
        viewBox="0 0 ${width} ${height}"
        style="font-size: 1.5px; font-family: monospace"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <pattern
            id="characters"
            width="3.55"
            height="5.95"
            patternUnits="userSpaceOnUse"
            x="12.45"
            y="12.55"
          >
            <rect width="2.95" height="5.55" fill-opacity="0.05" />
          </pattern>
          <pattern id="pins" width="2.54" height="3.255" patternUnits="userSpaceOnUse" y="1.1">
            <path
              fill="#92926d"
              d="M0,0.55c0,0 0.21,-0.52 0.87,-0.52 0.67,0 0.81,0.51 0.81,0.51v1.81h-1.869z"
            />
            <circle r="0.45" cx="0.827" cy="0.9" color="black" />
          </pattern>
        </defs>
        <rect width="${width}" height="${height}" fill="#087f45" />
        <rect x="4.95" y="5.7" width="${panelWidth + 15}" height="${panelHeight + 13.7}" />
        <rect
          x="7.55"
          y="10.3"
          width="${panelWidth + 9.8}"
          height="${panelHeight + 4.5}"
          rx="1.5"
          ry="1.5"
          fill="${actualBgColor}"
        />
        <rect
          x="7.55"
          y="10.3"
          width="${panelWidth + 9.8}"
          height="${panelHeight + 4.5}"
          rx="1.5"
          ry="1.5"
          opacity="${darken}"
        />
        ${pins === 'i2c' ? this.renderI2CPins() : null}
        ${pins === 'full' ? this.renderPins(panelHeight) : null}
        <rect
          x="12.45"
          y="12.55"
          width="${panelWidth}"
          height="${panelHeight}"
          fill="url(#characters)"
        />
        <path d="${this.path(characters)}" transform="translate(12.45, 12.55)" fill="${color}" />
        ${this.renderCursor()}
      </svg>
    `;
    }
};
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "color", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "background", void 0);
__decorate([
    decorators_js_1.property({ type: Array })
], LCD1602Element.prototype, "characters", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "font", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "cursor", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "blink", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "cursorX", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "cursorY", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "backlight", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "pins", void 0);
__decorate([
    decorators_js_1.property()
], LCD1602Element.prototype, "text", null);
LCD1602Element = __decorate([
    decorators_js_1.customElement('wokwi-lcd1602')
], LCD1602Element);
exports.LCD1602Element = LCD1602Element;
